### ISSUE TITLE
cache the last data dir, the second time to train will be updated

### ISSUE THREAD

**Issue by figodeng** [ultralytics/ultralytics#13666]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### YOLOv8 Component

Train

### Bug

{'settings_version': '0.0.4', 'datasets_dir': './datasets/dataset_coco', 'weights_dir': 'weights', 'runs_dir': 'runs', 'uuid': '01f390beb77423aeaabd05ca005620f52db4283fdd039de671fc935e64e250e2', 'sync': True, 'api_key': '', 'openai_api_key': '', 'clearml': True, 'comet': True, 'dvc': True, 'hub': True, 'mlflow': True, 'neptune': True, 'raytune': True, 'tensorboard': True, 'wandb': True}

Transferred 355/355 items from pretrained weights
Freezing layer 'model.22.dfl.conv.weight'
train: Scanning /ultralytics-detection/datasets/coco128/labels/train.cache... 116 images, 2 backgrounds, 0 corrupt: 100%|██████████| 118/118 [00:00<?, ?it/s]
val: Scanning /ultralytics-detection/datasets/coco128/labels/test.cache... 10 images, 0 backgrounds, 0 corrupt: 100%|██████████| 10/10 [00:00<?, ?it/s]
WARNING ⚠️ 'name=model' automatically updated to 'name='.

### Environment

Ultralytics YOLOv8.2.31 🚀 Python-3.10.14 torch-2.3.1 CPU (Apple M1 Pro)
Setup complete ✅ (8 CPUs, 16.0 GB RAM, 333.0/460.4 GB disk)

OS                  macOS-14.4.1-arm64-arm-64bit
Environment         Darwin
Python              3.10.14
Install             pip
RAM                 16.00 GB
CPU                 Apple M1 Pro
CUDA                None

matplotlib          ✅ 3.9.0>=3.3.0
opencv-python       ✅ 4.10.0.82>=4.6.0
pillow              ✅ 10.3.0>=7.1.2
pyyaml              ✅ 6.0.1>=5.3.1
requests            ✅ 2.32.3>=2.23.0
scipy               ✅ 1.13.1>=1.4.1
torch               ✅ 2.3.1>=1.8.0
torchvision         ✅ 0.18.1>=0.9.0
tqdm                ✅ 4.66.4>=4.64.0
psutil              ✅ 5.9.8
py-cpuinfo          ✅ 9.0.0
pandas              ✅ 2.2.2>=1.1.4
seaborn             ✅ 0.13.2>=0.11.0
ultralytics-thop    ✅ 0.2.8>=0.2.5

### Minimal Reproducible Example

from ultralytics.models.yolo.detect import DetectionTrainer
from ultralytics.utils import yaml_load, IterableSimpleNamespace
from ultralytics import settings

def train(data_dir, save_dir, epochs=3):
    # Update a setting
    # bug: cache the last data dir, the second time to train will be updated
    settings.update({'datasets_dir': data_dir})
    # Update config
    args = dict(model='./config/pretrained_models/yolov8n.pt',
                data='./config/coco8.yaml',
                save_dir=save_dir,
                epochs=epochs)
    config = yaml_load('./config/default.yaml')
    trainer = DetectionTrainer(cfg=IterableSimpleNamespace(**config), overrides=args)
    print(settings)
    trainer.train()

train(data_dir='./datasets/dataset_coco', save_dir='./model')

### Additional

train(data_dir='./datasets/dataset_coco', save_dir='./model')

set data_dir to another path 'X', still loading data from './datasets/dataset_coco' the first time, use 'X' when you run the second time

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by glenn-jocher** [comment]
@figodeng hello!

Thank you for your detailed bug report and for providing a minimal reproducible example. This is very helpful! 😊

From your description, it seems like the issue is related to the caching mechanism of the dataset directory, where the first training run uses the old directory and only updates to the new directory on the second run.

To help us investigate further, could you please confirm the following:
1. Are you using the latest versions of `torch` and `ultralytics`? If not, please update to the latest versions and try running your code again.
2. Could you provide any additional logs or error messages that appear during the first and second runs?

In the meantime, you might want to try explicitly clearing the cache before starting a new training session. Here’s a small modification to your code that might help:

```python
from ultralytics.models.yolo.detect import DetectionTrainer
from ultralytics.utils import yaml_load, IterableSimpleNamespace
from ultralytics import settings

def train(data_dir, save_dir, epochs=3):
    # Clear the cache
    settings.clear_cache()
    
    # Update settings
    settings.update({'datasets_dir': data_dir})
    
    # Update config
    args = dict(model='./config/pretrained_models/yolov8n.pt',
                data='./config/coco8.yaml',
                save_dir=save_dir,
                epochs=epochs)
    config = yaml_load('./config/default.yaml')
    trainer = DetectionTrainer(cfg=IterableSimpleNamespace(**config), overrides=args)
    print(settings)
    trainer.train()

train(data_dir='./datasets/dataset_coco', save_dir='./model')
```

This should ensure that the cache is cleared and the new dataset directory is used immediately.

Let me know if this helps or if you have any further questions!
